### PR TITLE
Initial development

### PR DESCRIPTION
**Pull request by ESKYoung**
# Summary

Initial development leveraging [PyGithub][pygithub] to programmatically call GitHub's REST APIv3. The main aim is to prevent accidental loss of access to organisation repositories when an individual repository admin leaves the organisation. 

This code provides a way to find all organisation repositories, contributors, and teams, and to set wholesale permissions of a team to one or more repositories. 

Main functions include:

1. Getting all repositories from a GitHub organisation;
2. Executing PyGithub [`github.Repository.Repository` class methods][pygithub-repository] on one or more repositories;
3. Extracting PyGitHub object attributes from a list or [`github.PaginatedList.PaginatedList`][pygithub-pagination]; and
4. Adding a team to one of more repositories (if it does not already have access), whilst concurrently setting permissions levels.

Tests are written in pytest, and have full code coverage. Documentation has been added, including example Juypter notebooks in the `notebooks` folder.

# Review request

Grateful if the reviewer could check that the functions all work as expected (see the [comment below](https://github.com/ukgovdatascience/github_organisation_administration/pull/1#issuecomment-755274756) on adding and setting team permissions). Any comments on improvements that should be implemented would also be welcome.

I've tried to maintain security by enforcing use of a separate JSON file to store GitHub credentials (username and personal access token), but would be grateful for comments/suggestions!

# Checklists

This pull/merge request meets the following requirements:

- [x] Code runs
- [x] Developments are **secure** and [**ethical**][data-ethics-framework]
- [x] Developments follow the AQA plan (see `docs/aqa/aqa_plan.md`)
- [x] You have made proportionate checks that the code works correctly
- [x] Test suite passes
- [x] Assumptions, and caveats log updated (see `docs/aqa/assumptions_caveats.md`), if necessary
- [x] [Minimum usable documentation][agilemodeling] written in the `docs` folder, and in other Markdown files where
      relevant

[agilemodeling]: http://agilemodeling.com/essays/documentLate.htm
[data-ethics-framework]: https://www.gov.uk/government/publications/data-ethics-framework
[pygithub]: https://pygithub.readthedocs.io
[pygithub-pagination]: https://pygithub.readthedocs.io/en/latest/utilities.html#pagination
[pygithub-repository]: https://pygithub.readthedocs.io/en/latest/github_objects/Repository.html

### CODE REVIEW

**Comment by ESKYoung** [comment]
For clarity, I tested the adding and setting team permissions IRL by executing the `add_team_with_permissions_to_all_repositories` function on **just** the `govcookiecutter` repository. After ensuring my own GitHub username had admin rights (i.e. not through a team), I ran two tests:

1. Removing a team (if needed) from `govcookiecutter` manually, and then executing the function to give the team set privileges. Verifying the changes manually.
2. With an existing team on `govcookiecutter`, execute the function to give the team a different set of privileges. Verifying the changes manually.

Instructions on how to execute the function for single, named repository can be found in `notebooks/example_set_team_permission_for_all_organisation_repositories.ipynb` under the "Setting a team's permission for a specific repository name" subsection.

**I did not run this wholesale across the entire `ukgovdatascience` organisation** - predominately because it's dangerous without reviewing this code, and also because I am not an organisation owner so wouldn't have permissions to run this for many repositories.

All other functions are non-destructive, so can be tested in IRL without issue.

**Comment by chao-xian** [comment]
I haven't had a chance to read through the code/commits yet but just a thought on the JSON file holding credentials: It's not clear to me why we're not re-using the `.secrets` file to read in these values? I worry that these credentials would be accidentally committed in an unknown JSON file.

File: src/utils/parse_api_token.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        key_username: The username key in file.
+        key_token: The token key in file.
+
+    Returns:
+        A tuple of two elements, where the first element is the username, and the second element is the token.
+
+    """
+
+    # Read and parse the JSON file, and return only the key_username, and key_token keys in the file
+    with open(file, "r") as f:
```
Comment (chao-xian) [inline src/utils/parse_api_token.py]: It's not clear to me why we're not re-using the `.secrets` file to read in these values? I worry that these credentials would be accidentally committed in an unknown JSON file.

**Review by chao-xian — COMMENTED** [review]
I haven't had a chance to read through the code/commits yet but just a thought on the JSON file holding credentials

File: src/utils/parse_api_token.py (hunk 1 above)
Comment (ESKYoung) [reply src/utils/parse_api_token.py]: That's a very good point @chao-xian! I think by trying to add extra security, I've introduced a flaw! 😬 

I've made changes to the code in commit [`33db318`](https://github.com/ukgovdatascience/github_organisation_administration/pull/1/commits/33db31875d366a26a677c27fbd721e8eae295769) to remove the `parse_api_token`, and all its references, and revised the documentation to state that the API token should be added as an environment variable in the `.secrets` file. Let me know what you think!

**Review by oscarwyatt — APPROVED** [review]
Super work!
Had a close look and I don't have any comments. I guess the only thing I'd say is that if anything, your testing is probably _too_ thorough! To my mind, testing the parallelizing process probably isn't necessary, as it can be verified by someone developing it and it isn't tied up to the "business" logic - but I certainly wouldn't insist you remove it!

Sterling work :+1: :100:

**Comment by ESKYoung** [comment]
Thanks for your review @oscarwyatt, and your helpful comments @chao-xian!

**Pull request merged**